otherlibs/dune-rpc/private/dune_rpc_private.ml

```diff
@@ -575,10 +266,16 @@ module Client = struct
     type t =
       { chan : Chan.t
       ; requests : (Id.t, Response.t Fiber.Ivar.t) Table.t
-      ; on_notification : Call.t -> unit Fiber.t
       ; initialize : Initialize.Request.t
       ; mutable next_id : int
       ; mutable running : bool
+      ; mutable handler_initialized : bool
+      ; (* We need this field to be an Ivar to ensure that any typed
+           communications are correctly versioned. The contract of the [Fiber]
+           interface ensures that this will be filled before any user code is
+           run. *)
```

Comment:
Can't we have `connect` wait for this handler to be determined before passing this to the user?

Reply:
It already does this (see line 698) -- we don't call `f` until the handler has been initialized, unless I'm misunderstanding you?

Reply:
No, you understand me correctly. I wasn't paying close attention. I guess my motivation for this is that I'd prefer if the negotiation did not return a fiber if it doesn't have to. I suppose a few additional binds aren't going to matter. We can leave things as is if you prefer.

Reply:
Yeah, it's a bit annoying to need to bind on the handler. I'm open to other ideas. We could use a regular ref/Fdecl/etc, but I thought that might be prone to races.

Reply:
It's fine to me as well. I'll cc @jeremiedimino to be aware just in case, as he's usually against all the extra closures.